### PR TITLE
bpo-35685: Add examples of unittest.mock.patch.dict usage

### PR DESCRIPTION
**Pull request by eamanu**
Add examples of the use of mock.path.dict with decorator and in class. Currently there are examples only with the use on context manager.

https://bugs.python.org/issue35685

### CODE REVIEW

**Comment by eamanu** [comment]
ping @terryjreedy :-) could you review it? thanks

File: Doc/library/unittest.mock.rst [hunk 1]
```diff
@@ -1308,9 +1308,23 @@ patch.dict
...
-    :func:`patch.dict` can be used as a context manager, decorator or class
-    decorator. When used as a class decorator :func:`patch.dict` honours
-    ``patch.TEST_PREFIX`` for choosing which methods to wrap.
+:func:`patch.dict` can be used as a context manager, decorator or class
+decorator. When used as a class decorator :func:`patch.dict` honours
+``patch.TEST_PREFIX`` for choosing which methods to wrap.
+
+    >>> foo = {}
+    >>> @patch.dict(foo, {'newkey': 'newvalue'})
+    >>> def test():
```
Comment (tirkarthi) [inline Doc/library/unittest.mock.rst]: Doctest fails. I think this needs to be 

```
>>> @patch.dict(foo, {'newkey': 'newvalue'})
... def test():
...     assert foo == {'newkey': 'newvalue'}
```

File: Doc/library/unittest.mock.rst [hunk 2]
```diff
@@ -1308,9 +1308,23 @@ patch.dict
...
+    ...     assert foo == {'newkey': 'newvalue'}
+    >>> test()
+    >>> assert foo == {}
+
+    >>> import os
+    >>> import unittest
+    >>> @patch.dict('os.environ', {'newkey': 'newvalue'}):
+    >>> class TestSample(unittest.TestCase):
+    ...     def test_sample(self):
+    ...         self.assertEqual(os.environ['newkey'], 'newvalue')
```
Comment (tirkarthi) [inline Doc/library/unittest.mock.rst]: CI failure with space in the end after ) .

**Comment by terryjreedy** [comment]
Tracker issues are more visible than PRs and more likely to get the attention of a core dev who can review and merge.  Michael Foord is the listed expert for unittest.mock, but does not have a listed GitHub account and therefore is inactive.  I am not an expert on either unittest or rst and have a backlog of IDLE PRs to review, so I will have to pass.

**Comment by eamanu** [comment]
> Tracker issues are more visible than PRs and more likely to get the attention of a core dev who can review and merge. Michael Foord is the listed expert for unittest.mock, but does not have a listed GitHub account and therefore is inactive. I am not an expert on either unittest or rst and have a backlog of IDLE PRs to review, so I will have to pass.

Thanks for the response. I will open a issue on bugs.python

File: Doc/library/unittest.mock.rst [hunk 3]
```diff
@@ -1308,9 +1308,23 @@ patch.dict
...
     values in the dictionary.
 
-    :func:`patch.dict` can be used as a context manager, decorator or class
-    decorator. When used as a class decorator :func:`patch.dict` honours
-    ``patch.TEST_PREFIX`` for choosing which methods to wrap.
+:func:`patch.dict` can be used as a context manager, decorator or class
+decorator. When used as a class decorator :func:`patch.dict` honours
+``patch.TEST_PREFIX`` for choosing which methods to wrap.
+
+    >>> foo = {}
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst:1583]: What do you think about moving this example before the "When used as a class decorator :func:`patch.dict` honours [...]" part? Seeing this example after reading a sentence about using `patch.dict()` as a class decorator seems confusing to me :)

File: Doc/library/unittest.mock.rst [hunk 4]
```diff
@@ -1308,9 +1308,23 @@ patch.dict
...
+    ...     assert foo == {'newkey': 'newvalue'}
+    >>> test()
+    >>> assert foo == {}
+
+    >>> import os
+    >>> import unittest
+    >>> @patch.dict('os.environ', {'newkey': 'newvalue'})
+    ... class TestSample(unittest.TestCase):
+    ...     def test_sample(self):
+    ...         self.assertEqual(os.environ['newkey'], 'newvalue')
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst:1599]: Could you please add another method to demonstrate that methods starting with other than `patch.TEST_PREFIX` won't be wrapped by `patch.dict()`?

File: Doc/library/unittest.mock.rst [hunk 5]
```diff
@@ -1308,9 +1308,23 @@ patch.dict
     :func:`patch.dict` can also be called with arbitrary keyword arguments to set
     values in the dictionary.
 
-    :func:`patch.dict` can be used as a context manager, decorator or class
-    decorator. When used as a class decorator :func:`patch.dict` honours
-    ``patch.TEST_PREFIX`` for choosing which methods to wrap.
+:func:`patch.dict` can be used as a context manager, decorator or class
+decorator. When used as a class decorator :func:`patch.dict` honours
+``patch.TEST_PREFIX`` for choosing which methods to wrap.
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst]: I know this is just a copy & paste, but it would be nice to make the `patch.dict()` documentation self-contained and mention that the default value of `patch.TEST_PREFIX` is `test`. Or you could just link to https://docs.python.org/3/library/unittest.mock.html#test-prefix

Feel free to change other `patch.*` docs.

**Review by berkerpeksag — CHANGES_REQUESTED** [review]
Thank you for wanting to improve our documentation! :) I just left some comments.

File: Doc/library/unittest.mock.rst (hunk 3 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst:1583]: Great, it's a good idea. I didn't want break the existing paragraphs. I will change.

File: Doc/library/unittest.mock.rst (hunk 5 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst]: Ok, I will copy & paste I think that is more useful have this in the same paragraph.
 
> Feel free to change other `patch.*` docs.

Ok. I focused on patch.dict because I was working on https://github.com/python/cpython/pull/11437

After finish this PR with patch.dict, I will review whole patch.*  :-)

File: Doc/library/unittest.mock.rst (hunk 4 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst:1599]: Ok!

**Comment by eamanu** [comment]
I have made the requested changes; please review again

File: Doc/library/unittest.mock.rst [hunk 6]
```diff
@@ -1308,9 +1308,32 @@ patch.dict
     :func:`patch.dict` can also be called with arbitrary keyword arguments to set
     values in the dictionary.
 
-    :func:`patch.dict` can be used as a context manager, decorator or class
-    decorator. When used as a class decorator :func:`patch.dict` honours
-    ``patch.TEST_PREFIX`` for choosing which methods to wrap.
+:func:`patch.dict` can be used as a context manager, decorator or class
+decorator.
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst]: IMO, it would be better if we end the sentence with a colon.

File: Doc/library/unittest.mock.rst [hunk 7]
```diff
@@ -1308,9 +1308,32 @@ patch.dict
...
+    ... def test():
+    ...     assert foo == {'newkey': 'newvalue'}
+    >>> test()
+    >>> assert foo == {}
+
+When used as a class decorator :func:`patch.dict` honours
+``patch.TEST_PREFIX`` for choosing which methods to wrap. The patchers
+recognise methods that start with ``'test'`` as being test methods.
+If you want to use a different prefix for your test, you can inform the
+patchers of the different prefix by setting ``patch.TEST_PREFIX``.
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst]: IMO, it would be better if we end the sentence with a colon.

File: Doc/library/unittest.mock.rst [hunk 8]
```diff
@@ -1308,9 +1308,32 @@ patch.dict
...
+
+    >>> import os
+    >>> import unittest
+    >>> patch.TEST_PREFIX = 'foo'
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst]: I'd rather demonstrate the default behavior here. https://docs.python.org/3/library/unittest.mock.html#test-prefix already provides an example to show how to customize this behavior, so no need to repeat the same example here.

File: Doc/library/unittest.mock.rst [hunk 9]
```diff
@@ -1308,9 +1308,32 @@ patch.dict
...
+    >>> @patch.dict('os.environ', {'newkey': 'newvalue'})
+    ... class TestSample(unittest.TestCase):
+    ...     def foo_sample(self):
+    ...         self.assertEqual(os.environ['newkey'], 'newvalue')
+    ...
+    ...     def test_sample2(self):
+    ...         self.assertNotIn('newkey', os.environ.keys())
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst]: I think the following example is closer to a real word use case:

```py
with self.assertRaises(KeyError):
    os.environ['foo']
```

File: Doc/library/unittest.mock.rst [hunk 10]
```diff
@@ -1308,9 +1308,32 @@ patch.dict
...
+    >>> assert foo == {}
+
+When used as a class decorator :func:`patch.dict` honours
+``patch.TEST_PREFIX`` for choosing which methods to wrap. The patchers
+recognise methods that start with ``'test'`` as being test methods.
+If you want to use a different prefix for your test, you can inform the
+patchers of the different prefix by setting ``patch.TEST_PREFIX``.
+
+    >>> import os
+    >>> import unittest
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst:1594]: Could you also import `patch` here so the example can easily be pasted and tested without needing any modifications from readers?

File: Doc/library/unittest.mock.rst (hunk 3 above)
Comment (berkerpeksag) [reply Doc/library/unittest.mock.rst:1583]: Could you also import `patch` here so the example can easily be pasted and tested without needing any modifications from readers?

File: Doc/library/unittest.mock.rst [hunk 11]
```diff
@@ -1308,9 +1308,32 @@ patch.dict
...
+    >>> foo = {}
+    >>> @patch.dict(foo, {'newkey': 'newvalue'})
+    ... def test():
+    ...     assert foo == {'newkey': 'newvalue'}
+    >>> test()
+    >>> assert foo == {}
+
+When used as a class decorator :func:`patch.dict` honours
+``patch.TEST_PREFIX`` for choosing which methods to wrap. The patchers
+recognise methods that start with ``'test'`` as being test methods.
```
Comment (berkerpeksag) [inline Doc/library/unittest.mock.rst]: Thank you, but IMO this version is a bit wordy. I was thinking something like the following:

```rst
When used as a class decorator :func:`patch.dict` honours ``patch.TEST_PREFIX``
(defaults to ``'test'``) for choosing which methods to wrap.
```

We can also refer people to https://docs.python.org/3/library/unittest.mock.html#test-prefix for more details on how to use the feature:

```rst
For more details about how to change the value of ``patch.TEST_PREFIX``,
see :ref:`name-of-the-link`.
```

What do you think?

**Review by berkerpeksag — DISMISSED** [review]

File: Doc/library/unittest.mock.rst (hunk 7 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst]: Ready

Comment (eamanu) [reply Doc/library/unittest.mock.rst]: Ready

File: Doc/library/unittest.mock.rst (hunk 6 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst]: ready

File: Doc/library/unittest.mock.rst (hunk 8 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst]: Ok. I divide this paragraph in two, One to show the sample of use of the decorator on class. And the other has a link to patch.TEST_PREFIX sample.

File: Doc/library/unittest.mock.rst (hunk 9 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst]: I remove this sample, because I link to patch.TEST_PREFIX sample

File: Doc/library/unittest.mock.rst (hunk 10 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst:1594]: Ready

File: Doc/library/unittest.mock.rst (hunk 11 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst]: Ready! I added it

**Comment by eamanu** [comment]
I have made the requested changes; please review again

**Comment by eamanu** [comment]
ping @berkerpeksag :-)

File: Doc/library/unittest.mock.rst [hunk 12]
```diff
@@ -1308,9 +1308,31 @@ patch.dict
     :func:`patch.dict` can also be called with arbitrary keyword arguments to set
     values in the dictionary.
 
-    :func:`patch.dict` can be used as a context manager, decorator or class
-    decorator. When used as a class decorator :func:`patch.dict` honours
-    ``patch.TEST_PREFIX`` for choosing which methods to wrap.
+:func:`patch.dict` can be used as a context manager, decorator or class
+decorator:
+
+    >>> import patch
```
Comment (JulienPalard) [inline Doc/library/unittest.mock.rst]: You don't need to `import patch` (which would be `from unittest.mock import patch`), because it's already imported in the `testsetup` (top of the file). Currently it breaks the doctests.

**Review by JulienPalard — CHANGES_REQUESTED** [review]

File: Doc/library/unittest.mock.rst (hunk 12 above)
Comment (eamanu) [reply Doc/library/unittest.mock.rst]: Hello @JulienPalard I put this from the coment of @berkerpeksag https://github.com/python/cpython/pull/11456/files#r246624984.

But yes, this is broken the test. I'll remove

**Comment by eamanu** [comment]
I have made the requested changes; please review again

**Review by JulienPalard — APPROVED** [review]

**Review by berkerpeksag — APPROVED** [review]
